### PR TITLE
Upgrade to new resin-request

### PR DESCRIPTION
**Pull request by pimterry**
Only spotted this module in my resin-sdk PR just before I left for holidays, but it needs updating too. Main change here is updating the tests to work with the new resin-request, since the existing global fetch stub no longer works.

This module does manage its dependencies differently to what I've done in [resin-register-device](https://github.com/resin-io-modules/resin-register-device/pull/28). We should make these consistent before this or that gets merged.

Currently this module takes `dataDirectory` as an option, and builds its own resin-request and resin-token instances, passing the dataDirectory in itself. `resin-register-device` meanwhile takes a prebuilt resin-request, and just uses it - the equivalent here would be to pass in working `resin-token` and `resin-request` instances directly.

Feels nicer to inject them imo. It would mean this module isn't coupled to the resin-request or resin-token constructor arguments, and lets you share resin-request and token instances directly (although I'm not sure that makes much runtime difference, it feels cleaner).

Theoretically it would also give us the option of unit testing this with a stub `resin-request` [as in resin-register-device](https://github.com/resin-io-modules/resin-register-device/blob/415269d65ed9619ba91997323b20ad79e5a8e6ff/tests/register.spec.coffee#L23), rather than integration testing it and stubbing fetch (which is why this currently has to duplicate the fetch-mock workarounds from `resin-request`). 'Theoretical' because while would be tidier, I think I prefer the more thorough test behaviour we have right now for this module, but it's nice to have the option.

### CODE REVIEW

**Comment by emirotin** [comment]
> Currently this module takes dataDirectory as an option, and builds its own resin-request and resin-token instances, passing the dataDirectory in itself. 

Every other module does the same, resin-register-device is kind of an exception I think.

This is a useful diagram: https://ns-pxnflxuoaj.now.sh/resin-sdk%20modules.html
And this entire thread https://github.com/resin-io/resin-sdk/issues/221 as well

**Comment by emirotin** [comment]
> npm : Using '^' to denote the start of a url is deprecated. Use 'begin:' instead

Causes the fail in windows.

**Comment by emirotin** [comment]
BTW we should upgrade to the latest pinejs-client and it will require another util method passed to it as far as I remember.

**Comment by emirotin** [comment]
After thinking more about the alternatives: I'm against cross-package injection because it breaks the abstraction. Each package is supposed to manage its dependencies and if the dependencies are factories instead of singletons it shouldn't make difference IMO.
I'd convert https://github.com/resin-io-modules/resin-register-device/pull/28 to work with self-instantiated resin-request too. If that needs adding `dataDirectory` option it makes sense for me. We're making breaking changes and major release of almost every module (because they're converted from singletons to factories everywhere), so it looks like not a big deal to me.

**Comment by pimterry** [comment]
I've updated the fetch-mock usage to the new format, which should make the tests pass. I'll open a separate PR for the pinejs-client upgrade in a minute, since that's (hopefully) an independent change.

**Comment by pimterry** [comment]
Whoops, no, I won't merge this once the tests are done, nobody's approved it yet. If somebody could take a look when they have a minute that'd be great.

**Review by emirotin — APPROVED** [review]
LGTM but duplicating the mocks everywhere is tedious and error-prone indeed. Can we abstract it inside of `resin-request`? Like having a separate `resin-request/lib/setup-tests` module that can be easily used in all other modules that need to mock fetch? To be clear: I'm fine with merging it right now,  but it'll propagate to more modules very soon.

**Comment by emirotin** [comment]
And yes, the client upgrade is a separate change, just mentioned it because it's better to implement before we merge the SDK PR IMO.

**Comment by emirotin** [comment]
BTW I can play with abstracting the fetch mocks myself, looks like an interesting task to me.

**Comment by pimterry** [comment]
Yes, the mock setup annoys me too. Feel free to abstract it, that'd be nice. It is notably going to get much simpler soon though - there's completed PR open to fix most of it (https://github.com/wheresrhys/fetch-mock/pull/160) that's just waiting to merge, and the rest goes away once we fix https://github.com/wheresrhys/fetch-mock/issues/158 (which you're also welcome to have a go at, I haven't touched it yet).

I think once that's sorted the end result that gets shared should hopefully be something like:

```coffeescript
fetchMock = require('fetch-mock').sandbox(Promise)

if IS_BROWSER
    fetchMock.setGlobals(require('fetch-ponyfill')({ Promise }))

require('resin-request/build/utils').fetch = fetchMock
```

**Comment by pimterry** [comment]
Maybe the above snippet is even a meaningful standalone module? Extremely tiny, but probably useful to others. E.g. `fetch-ponyfill-mock`: takes a Promise implementation, returns a fetchMock sandbox using fetch-ponyfill that will work everywhere. With that, the code we have in each module is just:

```coffeescript
require('resin-request/build/utils').fetch = require('fetch-ponyfill-mock')(Promise)
```

**Comment by emirotin** [comment]
Yes, sounds reasonable. The above should also probably return an instance of `fetchMock` to configure fetches?

BTW should we expose the `setFetch` method from the `resin-request` to abstract the implementation details (specifically the `build/utils` part)?

**Comment by pimterry** [comment]
> BTW should we expose the setFetch method from the resin-request to abstract the implementation details (specifically the build/utils part)?

Yes, given that we're now going to be stubbing out fetch all over the place, that definitely seems better than digging into utils everywhere. I'd really like to find a nice way to use `mockery` to mock fetch-ponyfill directly, but it's tricky to get working nicely on both the node and browser side. Either making that work, or adding a top-level method to tidy this up sounds good to me. We should mark it as internal somehow though, nobody should really be using this for normal usage.

**Comment by emirotin** [comment]
But it still would be a legit usage (well what if they have their own fetch implementation they want for performance reasons or what else).
Anyway, let me file an issue there and have a look.

**Pull request merged**
